pkg/storage/ingest/partition_offset_client.go

```diff
@@ -0,0 +1,194 @@
+type partitionOffsetClient struct {
+	client *kgo.Client
+	admin  *kadm.Client
+	logger log.Logger
+	topic  string
+
+	// Metrics.
+	lastProducedOffsetRequestsTotal   *prometheus.CounterVec
+	lastProducedOffsetFailuresTotal   *prometheus.CounterVec
+	lastProducedOffsetLatency         *prometheus.HistogramVec
+	partitionStartOffsetRequestsTotal *prometheus.CounterVec
+	partitionStartOffsetFailuresTotal *prometheus.CounterVec
+	partitionStartOffsetLatency       *prometheus.HistogramVec
+}
+
+func newPartitionOffsetClient(client *kgo.Client, topic string, reg prometheus.Registerer, logger log.Logger) *partitionOffsetClient {
+	return &partitionOffsetClient{
+		client: client,
+		admin:  kadm.NewClient(client),
+		logger: logger,
+		topic:  topic,
+
+		lastProducedOffsetRequestsTotal: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "cortex_ingest_storage_reader_last_produced_offset_requests_total",
+			Help: "Total number of requests issued to get the last produced offset.",
+		}, []string{"partition"}),
+		lastProducedOffsetFailuresTotal: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "cortex_ingest_storage_reader_last_produced_offset_failures_total",
+			Help: "Total number of failed requests to get the last produced offset.",
+		}, []string{"partition"}),
+		lastProducedOffsetLatency: promauto.With(reg).NewHistogramVec(prometheus.HistogramOpts{
+			Name:                            "cortex_ingest_storage_reader_last_produced_offset_request_duration_seconds",
+			Help:                            "The duration of requests to fetch the last produced offset of a given partition.",
+			NativeHistogramBucketFactor:     1.1,
+			NativeHistogramMaxBucketNumber:  100,
+			NativeHistogramMinResetDuration: 1 * time.Hour,
+			Buckets:                         prometheus.DefBuckets,
+		}, []string{"partition"}),
+
+		partitionStartOffsetRequestsTotal: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "cortex_ingest_storage_reader_partition_start_offset_requests_total",
+			Help: "Total number of requests issued to get the partition start offset.",
+		}, []string{"partition"}),
+		partitionStartOffsetFailuresTotal: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "cortex_ingest_storage_reader_partition_start_offset_failures_total",
+			Help: "Total number of failed requests to get the partition start offset.",
+		}, []string{"partition"}),
+		partitionStartOffsetLatency: promauto.With(reg).NewHistogramVec(prometheus.HistogramOpts{
+			Name:                            "cortex_ingest_storage_reader_partition_start_offset_request_duration_seconds",
+			Help:                            "The duration of requests to fetch the start offset of a given partition.",
+			NativeHistogramBucketFactor:     1.1,
+			NativeHistogramMaxBucketNumber:  100,
+			NativeHistogramMinResetDuration: 1 * time.Hour,
+			Buckets:                         prometheus.DefBuckets,
+		}, []string{"partition"}),
```

Comment:
here we'd get one histogram per query-frontend per partition. Previously it would be only one per partition. Do we need this granularity? I'm not sure why we added it in the first place

the same applies to the rest of the metrics. Can we just remove the `partition` label?

Reply:
>here we'd get one histogram per query-frontend per partition

No, we don't. The query-frontends will request "all partitions" and my plan is to to use a special label value for that (`all`).

Reply:
ok, sounds good